RFC 0002 - MT Execution Contexts (#2)
crystal-lang/rfcs: 0002-execution-contexts.md

- Issue: -

# Summary

Reinvent MT support in with different Execution Contexts.

Note: this proposal takes inspiration from Go and Kotlin.

# Motivation (analysis of current situation)

## Terminology

- **Thread**: program whose execution is controlled by the operating system ([wikipedia](https://en.wikipedia.org/wiki/Thread_(computing))).
- **Fiber**: unit of works or “functions whose execution you can pause” ([wikipedia](https://en.wikipedia.org/wiki/Coroutine)), we can run multiple fibers per thread.
- **Scheduler**: manages fibers’ executions inside the program (controlled by Crystal), unlike threads that are scheduled by the operating system (outside of the accessibility of the program).
- **Event Loop**: an abstraction to wait on specific events, such as timers (e.g. wait until a certain amount of time has passed) or IO (e.g. wait for an socket to be readable or writable).
- **Multi-threaded (MT)**: the environment uses multiple threads, fibers can run both concurrently and in parallel on different CPU cores.
- **Single-threaded (ST)**: the environment uses only one thread, fibers can never run in parallel but still run concurrently.

## Concurrency

> > I only quickly sum up the concurrency model. There are very interesting topics about concurrency to consider (e.g. structured concurrency, fiber cancellation), but this RFC focuses on parallelism.

Reviewer: **suggestion:** GitHub Flavoured Markdown supports an admonition syntax s: https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax#alerts

```suggestion
> [!NOTE]
> I only quickly sum up the concurrency model. There are very interesting topics about concurrency to consider (e.g. structured concurrency, fiber cancellation), but this RFC focuses on parallelism.
```

Once spawned a fiber shouldn’t _move_ to another execution context. For example on re-enqueue the fiber must be resumed into it’s execution context: context B enqueues waiting sender from context A. That being said, we could allow to _send_ a fiber to another context.

## Kinds of execution contexts

The following are the potential contexts that Crystal could implement in stdlib.

**Single Threaded Context**: fibers will never run in parallel, they can use simpler and faster synchronization primitives internally (no atomics, no thread safety) and still communicate with other contexts with the default thread-safe primitives; the drawback is that a blocking fiber will block the thread and all the other fibers.

**Multi Threaded Context**: fibers will run in parallel and may be resumed by any thread, the number of schedulers and threads can grow or shrink, schedulers may move to another thread (M:N schedulers:threads) and steal fibers from each others; the advantage is that fibers that can run should be able to run, as long as a thread is available (i.e. no more starving threads) and we can be shrink the number of schedulers;

> **Isolated Context**: only one fiber is allowed to run on a dedicated thread (e.g. `Gtk.main` or game loop), thus disabling concurrency on that thread; an issue is what happens if the fiber tries to spawn (raise? forward to default context?), triggers the event-loop (IO, timers), or tries to communicate with other contexts? An advantage is that the fiber can own a thread, without blocking other fibers.

Author: You're right. I previously updated the RFC to reflect that:

1. The event loop can run normally: it suspends the current fiber waiting for the event and that's it.
2. We can specify another context to spawn fibers into (can be the default context by default).

Precisions:

- The above list isn’t exclusive: there can be other contexts with different rules (for example MT without work stealing).
- Each context isn’t exclusive: an application can start as many contexts in parallel as it needs.

## The default execution context

> Crystal would start a MT execution context with work-stealing where fibers are spawned by default. The goal of this context is to provide an environment that should fit most use cases to freely take advantage of multiple CPU cores, without developers having to think much about it, outside of protecting concurrent accesses to a resource or, preferably, using channels to communicate.

Author: There are more and more cores (even on x86) so limiting to one by default feels more and more dated.

I think it's beneficial to the Crystal ecosystem (shards) to know that they live in a parallel world _by default_ and it's easier & quicker to notice parallel issues, instead of a dual "usually ST" vs "sometimes MT".

> Crystal would start a MT execution context with work-stealing where fibers are spawned by default. The goal of this context is to provide an environment that should fit most use cases to freely take advantage of multiple CPU cores, without developers having to think much about it, outside of protecting concurrent accesses to a resource or, preferably, using channels to communicate.

Reviewer: Yeah of course we cannot simply rollout a change to run Crystal with multiple threads by default. That would be considered a breaking change. Many libraries are not thread-safe and would break on this.

So I think this document describes the desired state we want to reach eventually. We'll have to put fruther thought into how to get there and design the migration path.

> Crystal would start a MT execution context with work-stealing where fibers are spawned by default. The goal of this context is to provide an environment that should fit most use cases to freely take advantage of multiple CPU cores, without developers having to think much about it, outside of protecting concurrent accesses to a resource or, preferably, using channels to communicate.

Author: Exactly; a RFC should reflect the _desired_ state.

3. We can create an MT execution context for CPU heavy algorithms, that would block the current thread (e.g. hashing passwords using BCrypt with a high cost), and let the operating system preempt the threads, so the default context running a webapp backend won't be blocked.

4. The crystal compiler doesn’t need MT during the parsing and semantic phases (for now); we could configure the default execution context to 1 thread only, then start another execution context for codegen with as many threads as CPUs, and spawn that many fibers into this context.

5. Different contexts could have different priorities, maybe allowing the operating system to allocate threads more effectively in big.LITTLE architectures.

# Reference-level explanation

An execution context shall provide:

- configuration (e.g. number of threads, …);
- methods to spawn, enqueue, yield and reschedule fibers within its premises;
- a scheduler to run the fibers (or many schedulers for a MT context);
- an event loop (IO & timers):

>   => this might be complex: I don’t think we can share a libevent across event bases? we already need to have a “thread local” libevent object for IO objects as well as for PCRE2 (though this is an optimization).

Author: We currently have one per scheduler. I was thinking about one per EC, but you may be right: one per process might be enough and better :thinking:

>   => this might be complex: I don’t think we can share a libevent across event bases? we already need to have a “thread local” libevent object for IO objects as well as for PCRE2 (though this is an optimization).

Author: Yeah, I'm wondering about that. We currently have one libevent base per thread, and that overcomplexifies IO::Evented with lots of objects to allocate in the HEAP for each IO and thread.

I'm probably being naive (though Go seems to do that) but maybe a global event loop wouldn't behave so badly? Even with the potential contention on adding an event to the libevent base, especially on machines with too many cores to count (e.g. [ARM Neoverse](https://en.wikipedia.org/wiki/ARM_Neoverse)).

>   => this might be complex: I don’t think we can share a libevent across event bases? we already need to have a “thread local” libevent object for IO objects as well as for PCRE2 (though this is an optimization).

Reviewer 2: On the other hand not having to synchronize everything and their mother make other things a lot less complex. It is quite liberating to not give a shit about what other threads are doing. 

FWIW, having one dedicated ring per thread is also how the makers of io_uring recommend multi thread usage.

>   => this might be complex: I don’t think we can share a libevent across event bases? we already need to have a “thread local” libevent object for IO objects as well as for PCRE2 (though this is an optimization).

Reviewer 2: Yes, storing it on the stack is a very useful technique for io_uring as well (to keep it alive during submission). I've made a lot of use of that and it helped a lot. There the completion event need to be handled somehow though - I stored it in the fiber itself until it was awakened and could process it, but there may be better techniques.

One good thing about the current implementation though is that it handles the thundering herd problem decently, as far as I've been able to see. That is, if multiple fibers are waiting for something only one of them will wake when something happens. With many listeners that may become something to keep track of.

>   => this might be complex: I don’t think we can share a libevent across event bases? we already need to have a “thread local” libevent object for IO objects as well as for PCRE2 (though this is an optimization).

Author: We shall move the "refactor the event loop" into a proper issue on https://github.com/crystal-lang/crystal/issues

It doesn't need a RFC as it's mostly internal implementation detail.

>   => this might be complex: I don’t think we can share a libevent across event bases? we already need to have a “thread local” libevent object for IO objects as well as for PCRE2 (though this is an optimization).

Author: I might be interesting to consider one event loop (EL) per execution context (EC) :thinking:

With a single EL per EC, a starving scheduler will run the EL and enqueue every resumable fibers for the EC.

With one EL per scheduler, a starving scheduler will run its own EL and push the fraction of fibers it happened to have, possibly flooding the EC queues and delaying other schedulers from running their own EL, and delaying resumable fibers from being resumed.

That may be a reason why Go has a single EL: it could be unfair otherwise.

Both languages have coroutines and parallelism built right into the language. Both declare that coroutines may be resumed by whatever thread, and both runtimes expose an environment where multiple threads can resume the coroutines.

**Go** provides one execution context: their MT optimized solution. Every coroutine runs within that global context.

**Kotlin**, on the other hand, provides a similar environment to Go by default, but allows to [create additional execution contexts](https://kotlinlang.org/docs/coroutine-context-and-dispatchers.html) with one or more threads. Coroutines then live inside their designated context, which may each be ST or MT.

The **Tokio** crate for **Rust** provides a single-threaded scheduler or a multi-threaded scheduler (with work stealing), but it must be chosen at compilation time. Apparently the MT scheduler is selected by default (to be verified).

# Unresolved questions

## Introduction of the feature

The feature will likely require both the `preview_mt` and `execution_contexts` flags, at least for the initial implementations.

> Since it is a breaking change from stable Crystal, unrelated to `preview_mt`, MT and execution contexts may only become default with a Crystal major release?

Author: Default EC doesn't have to be ST: it can be MT:1 and the app could grow it :thinking:

> Maybe MT could be limited to 1 thread by default, and `spawn(same_thread)` be deprecated, but what to do with `spawn(same_thread: true)`?

Reviewer 3: Pass it to the EC and let it raise if not compatible. The default single threaded context could support this but any MT context that does work stealing would raise. Deprecate the parameter and schedule it for removal in 2.0.

## Default context configuration

> This proposal doesn’t solve the inherent problem of: how can applications configure the default context at runtime (e.g. number of MT schedulers) since we create the context before the application’s main can start. 

Reviewer 3: This proposal supports creating multiple execution contexts, let the application configure it's own EC and start fibers in it if required. That allows all the complexity the app needs when configuring the context the application actually runs in, because it's initialized by the application. The root context does not have to be well-used by the application.

3. We can create an MT execution context for CPU heavy algorithms, that would block the current thread (e.g. hashing passwords using BCrypt with a high cost), and let the operating system preempt the threads, so the default context running a webapp backend won't be blocked.

4. The crystal compiler doesn’t need MT during the parsing and semantic phases (for now); we could configure the default execution context to 1 thread only, then start another execution context for codegen with as many threads as CPUs, and spawn that many fibers into this context.

5. Different contexts could have different priorities, maybe allowing the operating system to allocate threads more effectively in big.LITTLE architectures.

# Reference-level explanation

An execution context shall provide:

- configuration (e.g. number of threads, …);
- methods to spawn, enqueue, yield and reschedule fibers within its premises;
- a scheduler to run the fibers (or many schedulers for a MT context);

> - an event loop (IO & timers):

Reviewer 2: It probably needs to be mentioned that it needs to continue to work with channels and mutexes. This is *somewhat* straightforward today with how fibers are bound to a thread once executed, so the simple version is to just schedule it using the normal interfaces. But interfaces for scheduling are not necessarily the same between execution contexts as they are within it! 

For example, a basic work stealing scheduler can default to just enqueue a fiber in the executing scheduler/thread context, and let distribution between threads happen in other ways. But that doesn't work if the thing to wake up doesn't live in the same context - then it needs to be communicated somehow. And then the question is what to communicate it to.

Also somewhere it should probably be defined what happens if channel interaction happens in an isolated context (as defined above).

  # reference to the current MT scheduler (only present for MT contexts)
  property! execution_context_scheduler : ExecutionContext::Scheduler

  # reference to the currently running fiber (simpler access + support scenarios
  # where a whole scheduler is moved to another thread when a fiber has blocked
  # for too long: the fiber would still need to access `Fiber.current`).
  property! current_fiber : Fiber
end

class Fiber
  def self.current : Fiber
    Thread.current.current_fiber
  end

  def self.yield : Nil
    ExecutionContext.yield
  end

  property execution_context : ExecutionContext

  def initialize(@name : String?, @execution_context : ExecutionContext)
  end

  def enqueue : Nil
    @execution_context.enqueue(self)
  end

  def resume : Nil
    # can't call ExecutionContext#resume directly (it's protected)
    ExecutionContext.resume(self)
  end
end

def spawn(*, name : String?, execution_context : ExecutionContext = ExecutionContext.current, &block) : Fiber
  Fiber.new(name, execution_context, &block)
end

def sleep : Nil
  ExecutionContext.reschedule
end

def sleep(time : Time::Span) : Nil
  ExecutionContext.sleep(time)
end
```

And the proposed API:

```crystal

> abstract class ExecutionContext

Reviewer 2: What is the lifetime of an execution context? Do we need some way to signal to it that it can close down when the last fiber has finished?

Author: No difference from today: the default EC hijacks the main thread and doesn't run the scheduler's run loop on that thread, so when the main thread's main fiber terminates that terminates the main thread and thus the process.

EC still need explicit sync (e.g. channel, wait group, nurseries, ...).

The shutdown procedure I mentioned is about stopping an EC, that would ask the threads to stop (cooperative) but we might ask to stop while there are runnable fibers (or fibers that will be reenqueued) and those should be cancelled, and... that sounds like something for structured concurrency (hence future evolution).

Reviewer 2: > the default EC hijacks the main thread and doesn't run the scheduler's run loop on that thread

Yes, this works for the main thread, and depends on the OS to do cleanup. I was more thinking of the nested case when someone creates a new execution context. How would the API for actual usage look like? Perhaps the lifecycle for children fall out of that?

Author: > Perhaps the lifecycle for children fall out of that?

Yes, I think it should be it's own type. Maybe a wrapping Execution Context? which brings an interesting point to the API: allow wrapping.

For example my current experiment assigns an EC::Scheduler < EC as the current EC, so wrapping would need to wrap both the EC (public interface) and EC::Scheduler (thread local interface) to monitor fibers :thinking:

Author: I added a couple notes about the ability the wrap an EC to add further capabilities (e.g. monitoring child fibers).

  # the default execution context (always started)
  class_getter default = MultiThreaded.new("DEFAULT", minimum: 1, maximum: System.cpu_count.to_i)

  def self.current : ExecutionContext
    Thread.current.execution_context
  end

  # the following methods delegate to the current execution context, they expose
  # the otherwise protected instance methods which are only safe to call on the
  # current execution context:

  def self.reschedule : Nil
    current.reschedule
  end

  def self.sleep(time : Time::Span) : Nil
    current.sleep(time)
  end

  def self.yield : Nil
    current.yield
  end

  def self.resume(fiber : Fiber) : Nil
    current = self.current
    if fiber.execution_context == current
      current.resume(fiber)
    else
      raise "BUG"
    end
  end

>   # the following methods can be called from whatever context and must thrad
>   # safe (even ST):

Reviewer 2: ```suggestion
  # the following methods can be called from whatever context and must be thread
  # safe (even ST):
```

  abstract def spawn(name : String?, &block) : Fiber
  abstract def enqueue(fiber : Fiber) : Nil

  # the following accessors don’t have to be protected, but their implementation
  # must be thread safe (even ST):

  abstract def stack_pool : Fiber::StackPool
  abstract def event_loop : Crystal::EventLoop

  # the following methods must only be called on the current execution context,
  # otherwise we could resume or suspend a fiber on whatever context:

  abstract protected def reschedule : Nil
  abstract protected def sleep(time : Time::Span) : Nil
  abstract protected def yield : Nil
  abstract protected def resume(fiber : Fiber) : Nil

  class SingleThreaded < ExecutionContext
    def initialize(name : String)
      # todo: start one thread
    end
  end

  class MultiThreaded < ExecutionContext
    getter name : String

    def self.new(name : String, exactly : Int32)
      new(name, exactly, exactly)
    end

    def initialize(@name : String, @minimum : Int32, @maximum : Int32)
      # todo: start @minimum threads

>     end

Reviewer 2: Allowing a dynamic amount of threads requires more synchronization and complexity than having a static amount. While it  sounds nice to be able to adjust, it probably warrants its own separate class. Making certain all threads are in a waiting state before starting to actually queue stuff allows a bunch of simplifications with less mutexes and risks a lot fewer possible race conditions too.

Author: Enqueue doesn't have to need much sync. Go pushes to a bounded local queue (per scheduler) with overflow to a global queue; threads can be started at any time: when they reach the run loop they will grab a batch of fibers from the global queue or steal from another scheduler. Stopping ain't more complex, schedulers aren't tied to a specific thread, the thread detaches the scheduler and returns itself to the thread pool.

The complexity is more in _when_ to start / stop a thread.

Author: And that complexity pushes for it to become a "future evolution".

   - or start an additional, single threaded, execution context.

2. Drop the `same_thread` option on `spawn`, instead:
   - limit the default context to a single thread (disables parallelism in the default context);
   - or create a single-threaded execution context for the fibers that must live on the same thread.

3. Drop the `preview_mt` flag and make execution contexts the only compilation mode (at worst introduce `without_mt`):
   - one can limit the default context to a single thread to (disabling parallelism at runtime);
   - sync primitives must be thread-safe, because fibers running on different threads or in different execution contexts will need to communicate safely;
   - sync primitives must be optimized for best possible performance when there is no parallelism;
   - community maintained shards may propose alternative sync primitives when we don’t want thread-safety to squeeze some extra performance inside a single-threaded context.

> [!NOTE]
> The breaking changes can be postponed to Crystal 2 by making the default execution context to be ST, keep supporting `same_thread: true` for ST while MT would raise, and `same_thread: false` would be a NOOP.
>

> > A compilation flag can be introduced to change the default context to be MT in Crystal 1 (e.g. keep `-Dpreview_mt` but consider just `-Dmt`).

Reviewer 2: Yet another option is to simply flip the default for a couple of versions (like 2.x), while allowing opting out of mt and new semantics given that flag

Ideally, anybody could implement an execution context that suits their application, or wrap an existing execution context.

## Examples

1. We can have fully isolated single-threaded execution contexts, mimicking the current MT implementation.

2. We can create an execution context dedicated to handle the UI or game loop of an application, and keep the threads of the default context to handle calculations or requests, never impacting the responsiveness of the UI.

3. We can create an MT execution context for CPU heavy algorithms, that would block the current thread (e.g. hashing passwords using BCrypt with a high cost), and let the operating system preempt the threads, so the default context running a webapp backend won't be blocked when thousands of users try to login at the same time.

4. The crystal compiler doesn’t need MT during the parsing and semantic phases (for now); we could configure the default execution context to 1 thread only, then start another execution context for codegen with as many threads as CPUs, and spawn that many fibers into this context.

> 5. Different contexts could have different priorities and affinities, to allow the operating system to allocate threads more effectively in big.LITTLE architectures.

Reviewer: **nitpick:** "big.LITTLE" is just one particular instance of a heteogenous computing architecture, I think it's better to use the generic term here to include all kinds of forms.
Also it should probably be "efficient", not "effective".

```suggestion
5. Different contexts could have different priorities and affinities, to allow the operating system to allocate threads more efficiently in heterogenous computing architectures.
```

    Thread.current.current_fiber
  end

  def self.yield : Nil
    ::sleep(0.seconds)
  end

  property execution_context : ExecutionContext

  def initialize(@name : String?, @execution_context : ExecutionContext)
  end

  def enqueue : Nil
    @execution_context.enqueue(self)
  end

  def resume : Nil
    # can't call ExecutionContext#resume directly (it's protected)
    ExecutionContext.resume(self)
  end
end

def spawn(*, name : String?, execution_context : ExecutionContext = ExecutionContext.current, &block) : Fiber
  Fiber.new(name, execution_context, &block)
end

def sleep : Nil
  ExecutionContext.reschedule
end

def sleep(time : Time::Span) : Nil
  Fiber.current.resume_event.add(time)
  ExecutionContext.reschedule
end
```

And the proposed API:

```crystal
abstract class ExecutionContext
  # the default execution context (always started)
  class_getter default = MultiThreaded.new("DEFAULT", minimum: 1, maximum: System.cpu_count.to_i)

  def self.current : ExecutionContext
    Thread.current.execution_context
  end

  # the following methods delegate to the current execution context, they expose
  # the otherwise protected instance methods which are only safe to call on the

>   # current execution context:

Reviewer: **question:** This part of the API feels a bit itchy. The methods are exposed publicly, but have a strict prerequisite: "only safe to call on the current execution context". I believe this is void for `.reschedule` because it explicitly targets `current`. So the only weird kid is `.resume`.

I'm not sure if you've given much thought into these peculiarities or just tried to map features of the current API to the new model as directly as possible.
Perhaps we might want to reconsider the implementation of `Fiber#resume`.
The general meaning of this method is to stop executing the current fiber and give control to the targeted fiber.
If they share the same EC, the targeted fiber can (potentially) replace the current one directly. This should be the choice of the EC, though.
If they are in different ECs, the semantics could be reached by enqueing the target fiber and stopping the current one.
```cr
  # Resumes `fiber` in the execution context. Raises if the fiber
  # doesn't belong to the context.
  def self.resume(fiber : Fiber) : Nil
    if fiber.execution_context == current
      current.resume(fiber)
    else
      fiber.execution_context.enque(fiber)
      current.reschedule
    end
  end
  ```

Author: I didn't give it much thought because:

1. it's not used anywhere in the stdlib, and only one spec relies on it; I'd go for deprecating `Fiber#resume` and drop it in Crystal 2;
2. it's already not working correctly with the current MT. It even used to be happily resumed by any thread :shrug:

I only got it "fixed" recently by raising on thread mismatch. It's probably in 1.11

Reviewer: Sounds good. Let's add `@[Deprecated]` then to document this.
